sdk/identity/Azure.Identity/src/ServiceFabricManagedIdentitySource.cs

```diff
@@ -0,0 +1,93 @@
+        {
+            string identityEndpoint = EnvironmentVariables.IdentityEndpoint;
+            string identityHeader = EnvironmentVariables.IdentityHeader;
+            string identityServerThumbprint = EnvironmentVariables.IdentityServerThumbprint;
+
+            if (string.IsNullOrEmpty(identityEndpoint) || string.IsNullOrEmpty(identityHeader) || string.IsNullOrEmpty(identityServerThumbprint))
+            {
+                return default;
+            }
+
+            Uri endpointUri;
+            try
+            {
+                endpointUri = new Uri(identityEndpoint);
+            }
+            catch (FormatException ex)
+            {
+                throw new AuthenticationFailedException(IdentityEndpointInvalidUriError, ex);
+            }
+
+            var httpHandler = new HttpClientHandler();
+
+            httpHandler.ServerCertificateCustomValidationCallback = ValidateMsiServerCertificate;
+
+            var customSslHttpPipline = HttpPipelineBuilder.Build(new TokenCredentialOptions { Transport = new HttpClientTransport(httpHandler) });
+
+            var innerPipeline = new CredentialPipeline(pipeline.AuthorityHost, customSslHttpPipline, pipeline.Diagnostics);
+
+            return new ServiceFabricManagedIdentitySource(innerPipeline, endpointUri, identityHeader, clientId);
+        }
+
+        private ServiceFabricManagedIdentitySource(CredentialPipeline pipeline, Uri endpoint, string identityHeaderValue, string clientId) : base(pipeline)
+        {
+            _endpoint = endpoint;
+            _identityHeaderValue = identityHeaderValue;
+            _clientId = clientId;
+        }
+
+        protected override Request CreateRequest(string[] scopes)
+        {
+            // covert the scopes to a resource string
+            string resource = ScopeUtilities.ScopesToResource(scopes);
+
+            Request request = Pipeline.HttpPipeline.CreateRequest();
+
+            request.Method = RequestMethod.Get;
+            request.Headers.Add("X-IDENTITY-HEADER", _identityHeaderValue);
+            request.Uri.Reset(_endpoint);
+            request.Uri.AppendQuery("api-version", ServiceFabricMsiApiVersion);
+            request.Uri.AppendQuery("resource", resource);
+
+            if (!string.IsNullOrEmpty(_clientId))
+            {
+                request.Uri.AppendQuery("client_id", _clientId);
+            }
+
+            return request;
+        }
+
+        private static bool ValidateMsiServerCertificate(HttpRequestMessage message, X509Certificate2 cert, X509Chain certChain, SslPolicyErrors policyErrors)
+        {
+            // Do any additional validation here
+            if (policyErrors == SslPolicyErrors.None)
+            {
+                return true;
+            }
+
+            return 0 == string.Compare(cert.GetCertHashString(), EnvironmentVariables.IdentityServerThumbprint, StringComparison.OrdinalIgnoreCase);
```

Comment:
Does cert.GetCertHashString() return the well formatted thumbprint "XX:XX:XX:XX:XX...." or "XXXXXXXX...." ?
If its well formatted then we'll need to remove the ':' separator before comparing.

Reply:
No separators just the hex value.